packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1503,69 +1602,322 @@ class Navigator extends StatefulWidget {
+        return true;
+      }());
+      result.add(navigator._routeNamed<dynamic>(Navigator.defaultRouteName, arguments: null, allowNull: true));
       final List<String> routeParts = initialRouteName.split('/');
       if (initialRouteName.isNotEmpty) {
         String routeName = '';
         for (String part in routeParts) {
           routeName += '/$part';
-          plannedInitialRoutes.add(_routeNamed<dynamic>(routeName, allowNull: true, arguments: null));
+          assert(() {
+            debugRouteNames.add(routeName);
+            return true;
+          }());
+          result.add(navigator._routeNamed<dynamic>(routeName, arguments: null, allowNull: true));
         }
       }
-      if (plannedInitialRoutes.last == null) {
+      if (result.last == null) {
         assert(() {
           FlutterError.reportError(
             FlutterErrorDetails(
               exception:
-                'Could not navigate to initial route.\n'
+              'Could not navigate to initial route.\n'
                 'The requested route name was: "/$initialRouteName"\n'
                 'There was no corresponding route in the app, and therefore the initial route specified will be '
                 'ignored and "${Navigator.defaultRouteName}" will be used instead.'
             ),
           );
           return true;
         }());
-        push(_routeNamed<Object>(Navigator.defaultRouteName, arguments: null));
-      } else {
-        plannedInitialRoutes.where((Route<dynamic> route) => route != null).forEach(push);
+        result.clear();
       }
+      // Null route might be a result of gap in initialRouteName
+      //
+      // For example, routes = ['A', 'A/B/C'], and initialRouteName = 'A/B/C'
+      // This should result in result = ['A', null,'A/B/C'] where 'A/B' produces
+      // the null. In this case, we want to filter out the null and return
+      // result = ['A', 'A/B/C'].
+      result.removeWhere((Route<dynamic> route) => route == null);
+    } else if (initialRouteName != Navigator.defaultRouteName) {
+      // If initialRouteName wasn't '/', then we try to get it with allowNull:true, so that if that fails,
+      // we fall back to '/' (without allowNull:true, see below).
+      result.add(navigator._routeNamed<dynamic>(initialRouteName, arguments: null, allowNull: true));
+    }
+    if (result.isEmpty)
+      result.add(navigator._routeNamed<dynamic>(Navigator.defaultRouteName, arguments: null));
+    return result;
+  }
+
+  @override
+  NavigatorState createState() => NavigatorState();
+}
+
+// The _RouteLifecycle state machine (only goes down):
+//
+//                    [creation of a _RouteEntry]
+//                   /  |             |       |
+//                  /   |             |       |
+//                 /    |             |       |
+//                /     |             |       |
+//               /      |             |       |
+//      pushReplace   push*      initial*   replace*
```

Comment:
we should rename "initial" to "add" since in the future we will also want to add non-initial routes without animation.

Reply:
(In the design I also had to add a new "adding#" state to properly support this)